src/error.rs

```diff
@@ -17,109 +19,127 @@ struct ErrorImpl {
     cause: Option<Cause>,
 }
 
-#[derive(Debug, PartialEq)]
-pub(super) enum Kind {
+/// Represents the kind of the error.
+///
+/// This enum is non-exhaustive.
+#[non_exhaustive]
+pub enum Kind {
+    /// Error occured while parsing.
     Parse(Parse),
+    /// Error occured while executing user code.
     User(User),
     /// A message reached EOF, but is not complete.
-    IncompleteMessage,
+    IncompleteMessage(Sealed),
     /// A connection received a message (or bytes) when not waiting for one.
     #[cfg(feature = "http1")]
-    UnexpectedMessage,
+    UnexpectedMessage(Sealed),
     /// A pending item was dropped before ever being processed.
-    Canceled,
+    Canceled(Sealed),
     /// Indicates a channel (client or body sender) is closed.
-    ChannelClosed,
+    ChannelClosed(Sealed),
     /// An `io::Error` that occurred while trying to read or write to a network stream.
     #[cfg(any(feature = "http1", feature = "http2"))]
-    Io,
+    Io(Sealed),
     /// Error occurred while connecting.
-    Connect,
+    Connect(Sealed),
     /// Error creating a TcpListener.
     #[cfg(all(
         any(feature = "http1", feature = "http2"),
         feature = "tcp",
         feature = "server"
     ))]
-    Listen,
+    Listen(Sealed),
     /// Error accepting on an Incoming stream.
     #[cfg(any(feature = "http1", feature = "http2"))]
     #[cfg(feature = "server")]
-    Accept,
+    Accept(Sealed),
     /// Error while reading a body from connection.
     #[cfg(any(feature = "http1", feature = "http2", feature = "stream"))]
-    Body,
+    Body(Sealed),
     /// Error while writing a body to connection.
     #[cfg(any(feature = "http1", feature = "http2"))]
-    BodyWrite,
+    BodyWrite(Sealed),
     /// The body write was aborted.
-    BodyWriteAborted,
+    BodyWriteAborted(Sealed),
     /// Error calling AsyncWrite::shutdown()
     #[cfg(feature = "http1")]
-    Shutdown,
+    Shutdown(Sealed),
 
     /// A general error from h2.
     #[cfg(feature = "http2")]
-    Http2,
+    Http2(Sealed),
 }
 
-#[derive(Debug, PartialEq)]
-pub(super) enum Parse {
-    Method,
-    Version,
+/// Represents the kind of the parse error.
+///
+/// This enum is non-exhaustive.
+#[non_exhaustive]
+pub enum Parse {
+    /// Invalid HTTP method parsed.
+    Method(Sealed),
+    /// Invalid HTTP version parsed.
+    Version(Sealed),
+    /// Found HTTP/2 preface.
     #[cfg(feature = "http1")]
-    VersionH2,
-    Uri,
-    Header,
-    TooLarge,
-    Status,
+    H2Preface(Sealed),
+    /// Invalid URI.
+    Uri(Sealed),
+    /// Invalid HTTP header parsed.
+    Header(Sealed),
+    /// Header section is too large.
+    HeaderSectionTooLarge(Sealed),
+    /// Invalid HTTP status-code parsed.
+    Status(Sealed),
 }
 
-#[derive(Debug, PartialEq)]
-pub(super) enum User {
+/// Represents the kind of the user error.
+///
+/// This enum is non-exhaustive.
+#[non_exhaustive]
+pub enum User {
```

Comment:
I kinda think I wouldn't want to expose the user error types. At least not yet, I think they grew without too much planning. And some are specifically to give a better error message, but I'd want to be able to change the kind to something better if they exist.

Reply:
Now that I've mentioned wanting private variants, this is an error where I wish we had them. I'm not sure if all user errors should be at a top level `Kind::User(_)`, or if they should be split up. But with `Kind::User(_)` catching what's currently here, it'd be a behavior change to move them... That's my conundrum.

Reply:
If you have thoughts on how to rework `User` errors, I'm all ears cause I don't like them very much.

For example, I don't see what "the user" has to do with errors returned by `HttpBody::poll_data`. To me the user of Hyper is the author of the product that makes use of Hyper, not the eyeball who is using our product and sending us stuff that makes `poll_data` return an error.

For this reason, I very much want the various user variants to be exposed.

Reply:
My perception of user is a user of hyper. So the application developer who used hyper to do a thing.

It may also be true that some of the things I have classified as user errors aren't exactly. It's a little fuzzy.